### PR TITLE
[API][AddressBook] Protect addresses with new firewall

### PR DESCRIPTION
**Pull request by AdamKasp**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Address.xml [hunk 1]
```diff
@@ -12,7 +12,7 @@
            xsi:schemaLocation="https://api-platform.com/schema/metadata https://api-platform.com/schema/metadata/metadata-2.0.xsd"
 >
     <resource class="%sylius.model.address.class%" shortName="Address">
-        <attribute name="route_prefix">shop</attribute>
+        <attribute name="route_prefix">shop/account</attribute>
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Address.xml:15]: I'm not sure about this change. Do we serialize the whole address to orders or do we have IRI to fetch it on your own?

File: src/Sylius/Behat/Context/Api/Shop/AddressContext.php [hunk 2]
```diff
@@ -163,6 +169,14 @@ public function iSetTheAddressOfAsDefault(string $fullName): void
         $this->customerClient->update();
     }
 
+    /**
+     * @When /^I try to view details of (address belonging to "([^"]+)")$/
+     */
+    public function iTryToViewDetailsOfAddressBelongingTo(AddressInterface $address): void
+    {
+        $this->client->showByIri($this->iriConverter->getIriFromItem($address));
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/AddressContext.php:177]: I guess it should be:
```suggestion
        $this->addressClient->showByIri($this->iriConverter->getIriFromItem($address));
```

**Review by GSadee — COMMENTED** [review]
There should be also scenarios for inability to add and delete an address from address book by a visitor

File: src/Sylius/Behat/Context/Transform/AddressContext.php [hunk 3]
```diff
@@ -150,6 +150,7 @@ public function getByStreet($street)
 
     /**
      * @Transform /^address of "([^"]+)"$/
+     * @Transform /^address belonging to "([^"]+)"$/
```
Comment (arti0090) [inline src/Sylius/Behat/Context/Transform/AddressContext.php:153]: More correct form should be `address belongs to` 😄 🇬🇧

File: src/Sylius/Bundle/ApiBundle/DataProvider/AddressItemDataProvider.php [hunk 4]
```diff
@@ -54,7 +54,7 @@ public function getItem(string $resourceClass, $id, string $operationName = null
             in_array('ROLE_USER', $user->getRoles(), true) &&
```
Comment (arti0090) [inline src/Sylius/Bundle/ApiBundle/DataProvider/AddressItemDataProvider.php:54]: It is not part of this PR but maybe you could change something so `$user instanceof ShopUserInterface` won't be called 2 times? 🚀

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Address.xml (hunk 1 above)
Comment (AdamKasp) [reply src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Address.xml:15]: `src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Address.xml` in this file we are serializing all properties from the address in `checkout` and `orderShow`

File: src/Sylius/Bundle/ApiBundle/DataProvider/AddressItemDataProvider.php (hunk 4 above)
Comment (AdamKasp) [reply src/Sylius/Bundle/ApiBundle/DataProvider/AddressItemDataProvider.php:54]: maybe :D but like you said it isn't the scope of this PR :)

File: src/Sylius/Behat/Context/Api/Shop/AddressContext.php [hunk 5]
```diff
@@ -151,6 +158,14 @@ public function iDeleteTheAddress(string $fullName): void
         $this->addressClient->delete($id);
     }
 
+    /**
+     * @When /^I try to delete (address belongs to "([^"]+)")$/
+     */
+    public function iDeleteTheAddressBelongsTo(AddressInterface $address): void
+    {
+        $this->addressClient->delete($this->iriConverter->getIriFromItem($address));
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/AddressContext.php:166]: Shouldn't it looks like:
```suggestion
        $this->addressClient->delete($address->getId());
```

File: config/packages/security.yaml [hunk 6]
```diff
@@ -145,5 +147,6 @@ security:
 
         - { path: "%sylius.security.new_api_admin_regex%/.*", role: ROLE_API_ACCESS }
         - { path: "%sylius.security.new_api_route%/admin/authentication-token", role: IS_AUTHENTICATED_ANONYMOUSLY }
-        - { path: "%sylius.security.new_api_shop_regex%/.*", role: IS_AUTHENTICATED_ANONYMOUSLY }
+        - { path: "%sylius.security.new_api_user_account_regex%/.*", role: ROLE_USER }
         - { path: "%sylius.security.new_api_route%/shop/authentication-token", role: IS_AUTHENTICATED_ANONYMOUSLY }
+        - { path: "%sylius.security.new_api_shop_regex%/.*", role: IS_AUTHENTICATED_ANONYMOUSLY }
```
Comment (GSadee) [inline config/packages/security.yaml:152]: Why did you move this line to the end?

File: config/packages/security.yaml [hunk 7]
```diff
@@ -145,5 +147,6 @@ security:
 
         - { path: "%sylius.security.new_api_admin_regex%/.*", role: ROLE_API_ACCESS }
         - { path: "%sylius.security.new_api_route%/admin/authentication-token", role: IS_AUTHENTICATED_ANONYMOUSLY }
-        - { path: "%sylius.security.new_api_shop_regex%/.*", role: IS_AUTHENTICATED_ANONYMOUSLY }
+        - { path: "%sylius.security.new_api_user_account_regex%/.*", role: ROLE_USER }
```
Comment (GSadee) [inline config/packages/security.yaml:150]: Changes in `security.yaml` should be mentioned in UPGRADE file

File: features/account/customer_account/address_book/preventing_not_logged_user_from_operations_on_address_book.feature [hunk 8]
```diff
@@ -0,0 +1,27 @@
+@address_book
+Feature: Preventing not logged user from operations on the address book
+    In order to protect address book from unauthorised operation
+    As a Visitor
+    I want to not be able to operate on address book
```
Comment (GSadee) [inline features/account/customer_account/address_book/preventing_not_logged_user_from_operations_on_address_book.feature:5]:
```suggestion
    I want not to be able to operate on address book
```

File: src/Sylius/Behat/Context/Api/Shop/AddressContext.php [hunk 9]
```diff
@@ -306,6 +329,30 @@ public function iShouldBeNotifiedThatAddressHasBeenSetAsDefault(): void
         Assert::true($this->responseChecker->isUpdateSuccessful($this->addressClient->getLastResponse()));
     }
 
+    /**
+     * @Then I should not see any details of address
+     */
+    public function iShouldNotSeeAnyDetailsOfAddress(): void
+    {
+        Assert::same($this->responseChecker->getError($this->addressClient->getLastResponse()), 'JWT Token not found');
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/AddressContext.php:337]:
```suggestion
        Assert::true($this->responseChecker->hasAccessDenied($this->addressClient->getLastResponse()));
```

File: UPGRADE-1.9.md [hunk 10]
```diff
@@ -0,0 +1,16 @@
+# UPGRADE FROM `v1.8.X` TO `v1.9.0`
+
+1. Add new parameters, new access control and reorder it:
```
Comment (GSadee) [inline UPGRADE-1.9.md:3]:
```suggestion
1. Add new parameters, new access control configuration and reorder it:
```

File: src/Sylius/Behat/Context/Api/Shop/AddressContext.php [hunk 11]
```diff
@@ -306,6 +329,30 @@ public function iShouldBeNotifiedThatAddressHasBeenSetAsDefault(): void
...
+    }
+
+    /**
+     * @Then I should not be able to add it
+     */
+    public function iShouldNotBeAbleToDoIt(): void
+    {
+        Assert::false($this->responseChecker->isCreationSuccessful($this->addressClient->getLastResponse()));
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/AddressContext.php:345]: Maybe check here if the access has been denied

**Review by GSadee — APPROVED** [review]

File: src/Sylius/Behat/Context/Api/Shop/AddressContext.php [hunk 12]
```diff
@@ -306,6 +329,29 @@ public function iShouldBeNotifiedThatAddressHasBeenSetAsDefault(): void
         Assert::true($this->responseChecker->isUpdateSuccessful($this->addressClient->getLastResponse()));
     }
 
+    /**
+     * @Then I should not see any details of address
+     */
+    public function iShouldNotSeeAnyDetailsOfAddress(): void
+    {
+        Assert::true($this->responseChecker->hasAccessDenied($this->addressClient->getLastResponse()));    }
```
Comment (arti0090) [inline src/Sylius/Behat/Context/Api/Shop/AddressContext.php:337]: I would take `}` one line down 😨

**Review by arti0090 — APPROVED** [review]

**Comment by GSadee** [comment]
Thank you, Adam! :tada:
